scipy/optimize/_sda.py

```diff
@@ -0,0 +1,572 @@
+        do_ls = False
+        if self.K < 90 * len(self.state.current_location):
+            pls = np.exp(self.K * (self.state.ebest - self.state.current_energy
+                                   ) / self.temperature_step)
+            if pls >= self._rs.random_sample():
+                do_ls = True
+        # Global energy not improved, let's see what LS gives
+        # on the best Markov chain location
+        if self.not_improved_idx >= self.not_improved_max_idx:
+            do_ls = True
+        if do_ls:
+            e, x = self.ofw.local_search(self.xmin)
+            self.xmin = np.copy(x)
+            self.emin = e
+            self.not_improved_idx = 0
+            self.not_improved_max_idx = self.state.current_location.size
+            if e < self.state.ebest:
+                self.state.ebest = self.emin
+                self.state.xbest = np.copy(self.xmin)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+
+
+class ObjectiveFunWrapper(object):
+    """
+    Class used to wrap around the objective function in order to apply local
+    search and default gradient computation.
+    Default local minimizer is L-BFGS-B
+    """
+    def __init__(self, bounds, func, **kwargs):
+        self.func = func
+        self.nb_fun_call = 0
+        self.kwargs = kwargs
+        self.minimizer = minimize
+        self.fun_args = None
+        lu = list(zip(*bounds))
+        self.lower = np.array(lu[0])
+        self.upper = np.array(lu[1])
+        self.ls_max_iter = self.lower.size * 6
+
+        if self.ls_max_iter < 100:
+            self.ls_max_iter = 100
+        if self.ls_max_iter > 1000:
+            self.ls_max_iter = 1000
+
+        # By default, scipy L-BFGS-B is used with a custom 3 points gradient
+        # computation
+        else:
+            self.fun_args = ()
+        if not self.kwargs or 'method' not in self.kwargs:
+            self.kwargs['method'] = 'L-BFGS-B'
+            self.kwargs['options'] = {
+                'disp': None, 'maxls': 100, 'iprint': -1, 'gtol': 1e-06,
+                'eps': 1e-06,
+                'maxiter': self.ls_max_iter,
+                'maxcor': 10, 'maxfun': 15000
+            }
+            if 'jac' not in self.kwargs:
+                self.kwargs['jac'] = self.gradient
+            if 'bounds' not in self.kwargs:
+                self.kwargs['bounds'] = bounds
+        if 'args' in self.kwargs:
+            self.fun_args = self.kwargs.get('args')
+
+    def func_wrapper(self, x):
+        self.nb_fun_call += 1
+        return self.func(x, *self.fun_args)
+
+    def gradient(self, x):
+        g = approx_derivative(self.func_wrapper, x)
+        return g
+
+    def local_search(self, x, maxlsiter=None):
+        mres = self.minimizer(self.func_wrapper, x, **self.kwargs)
+        if not mres.success:
+            return BIG_VALUE, None
+        return (mres.fun, mres.x)
+
+
+class SDARunner(object):
```

Comment:
It seems to me that you actually don't need this class at all, since you could just run the code inside ``__init__``, ``search`` and ``result`` sequentially.

It is not very critical, but I think it may make your code simpler if you just move this blocks of code to inside ``sda`` function, since you are already running them sequentially.

Reply:
SDARunner class removed, and it actually make the code simpler.